keps/sig-node/2008-checkpoint-restore/README.md

```diff
@@ -0,0 +1,597 @@
+in a compressed (or uncompressed) tar archive.
+
+The default location for the kubelet checkpoints is currently below `--root-dir`
+and defaults to `/var/lib/kubelet/checkpoints`. This directory can be anywhere. A
+local directory or especially for cluster setup where Pods will be migrated, a
+network mounted storage.
+
+To work with all possible storage backends the only requirement on the
+checkpoint storage is that it can be used to store and retrieve the previously
+mentioned Pod checkpoint archive.
+
+Additionally there have been discussions to use the OCI image format to store
+checkpoint archives. With this approach it would be possible to push a
+checkpoint to a registry and also retrieve existing images from a registry.
+
+Another possibility is to use a checkpoint image server which can be used to
+stream checkpoints to and where they can be later retrieved from. This is,
+however, not part of the current KEP and pull requests. Just mentioned as
+additional information.
+
+The actual checkpointing does not require much additional memory. CRIU
+writes the pages directly to disk so that checkpointing should not be
+problematic concerning cgroup memory limits.
+
+### Hooks
+
+A possible use case is fast startup of JVMs by starting from a checkpoint and
+especially for the JVM use case we have been discussing the possibilities to
+have a way to tell the application in the container that it is about to be
+checkpointed. The JVM could drop certain caches to remove unnecessary
+information from the checkpoint or also remove secrets from memory so that
+these things are not included in the checkpoint.
+
+Once a container has been restored it would also be helpful to let the
+application know that it has been restored.
+
+### User Stories
+
+As this KEP is explicitly not about container migration the user stories
+are not mentioning container migration based user stories. If this would
+be helpful to understand this KEP better it can be easily added.
+
+#### Story 1
+
+To analyze unusual activities in a container, the container should
+be checkpointed without stopping the container or without the container
+knowing it was checkpointed. Using checkpointing it is possible to take
+a copy of a running container for forensic analysis. The container will
+continue to run without knowing a copy was created. This copy can then
+be restored in another (sandboxed) environment in the context of another
+container engine for detailed analysis of a possible attack.
+
+#### Story 2
+
+Although containers are supposed to be stateless there are still containers
+which either require some time to startup or which have data cached.
+If a system is rebooted to update the kernel these containers can be
+checkpointed. Once the system has been rebooted into the new kernel the
+container can be restored and continue from its previous memory state
+without the need to wait for long start up times or to reload cached data
```

Comment:
It is more similar to snapshots in database, so the start up time (loading from image vs loading from disk) may not be the main benefit of this story. For in-memory cached data, which may keep changing over time, checkpoint-restore can help save the work of writing its own backing up into disk and recovering logics.